File: medicine_app/planning.py
from __future__ import annotations

import sqlite3
from datetime import date, datetime
from typing import Callable

from .safety import APP_TIMEZONE


def clock_sort_key(value: str) -> tuple[int, int]:
    """Parse a validated clock value numerically, including legacy H:MM rows."""
    hour_text, minute_text = value.split(":", 1)
    hour, minute = int(hour_text), int(minute_text)
    if not 0 <= hour <= 23 or not 0 <= minute <= 59:
        raise ValueError("schedule time must be HH:MM")
    return hour, minute


def medication_course_progress(medication: dict, target: date) -> dict | None:
    """Return calendar progress; active remaining days exclude the target day."""
    if not medication.get("start_date") or not medication.get("end_date"):
        return None
    start = date.fromisoformat(medication["start_date"])
    end = date.fromisoformat(medication["end_date"])
    total_days = (end - start).days + 1
    if total_days <= 0:
        raise ValueError("medication end_date must be on or after start_date")
    if target < start:
        status, current_day, remaining_days, progress_percent = "upcoming", 0, total_days, 0
    elif target > end:
        status, current_day, remaining_days, progress_percent = "completed", total_days, 0, 100
    else:
        current_day = (target - start).days + 1
        remaining_days = (end - target).days
        status = "active"
        progress_percent = round(current_day * 100 / total_days)
    return {
        "status": status,
        "total_days": total_days,
        "current_day": current_day,
        "remaining_days": remaining_days,
        "progress_percent": progress_percent,
    }


def sort_medications_by_time(medications: list[dict], target: date) -> list[dict]:
    """Annotate courses and sort explicit daily times before unscheduled medication."""
    annotated: list[dict] = []
    for medication in medications:
        item = dict(medication)
        item["course_progress"] = medication_course_progress(item, target)
        annotated.append(item)

    def key(item: dict) -> tuple[int, str]:
        times = [schedule["time_of_day"] for schedule in item.get("schedules") or []]
        if not times:
            return 1, ""
        hour, minute = min(clock_sort_key(value) for value in times)
        return 0, f"{hour:02d}:{minute:02d}"

    # Python's stable sort preserves the database list order for equal times.
    return sorted(annotated, key=key)


def medication_applies_on(medication: dict, target: date) -> bool:
    if not medication.get("active"):
        return False
    start = date.fromisoformat(medication["start_date"]) if medication.get("start_date") else None
    end = date.fromisoformat(medication["end_date"]) if medication.get("end_date") else None
    if start and target < start:
        return False
    if end and target > end:
        return False
    return True


def _desired_for_medication(medication: dict, target: date) -> list[dict]:
    dose_text = medication.get("dosage_text")
    schedules = sorted(
        medication.get("schedules") or [],
        key=lambda schedule: clock_sort_key(schedule["time_of_day"]),
    )
    if schedules:
        return [
            {
                "medication_id": medication["id"],
                "scheduled_date": target.isoformat(),
                "schedule_key": f"slot:{index}",
                "scheduled_time": schedule["time_of_day"],
                "slot_label": None,
                "dose_text": schedule.get("dose_text") or dose_text,
            }
            for index, schedule in enumerate(schedules, 1)
        ]

    frequency = medication.get("frequency_per_day")
    if frequency:
        return [
            {
                "medication_id": medication["id"],
                "scheduled_date": target.isoformat(),
                "schedule_key": f"slot:{index}",
                "scheduled_time": None,
                "slot_label": f"{index}회차",
                "dose_text": dose_text,
            }
            for index in range(1, int(frequency) + 1)
        ]
    return []


def materialize_daily_plan(
    con: sqlite3.Connection,
    person_id: str,
    medications: list[dict],
    target: date,
    uuid_factory: Callable[[], str],
) -> dict:
    applicable = [med for med in medications if medication_applies_on(med, target)]
    prn = [med for med in applicable if med.get("as_needed")]
    scheduled = [med for med in applicable if not med.get("as_needed")]
    unscheduled = [med for med in scheduled if not med.get("schedules") and not med.get("frequency_per_day")]

    desired = [item for med in scheduled for item in _desired_for_medication(med, target)]
    desired_keys = {(item["medication_id"], item["schedule_key"]) for item in desired}

    existing = con.execute(
        """
        SELECT id, medication_id, schedule_key, status, scheduled_time, slot_label, dose_text
        FROM dose_instances WHERE person_id=? AND scheduled_date=?
        """,
        (person_id, target.isoformat()),
    ).fetchall()
    existing_by_key = {
        (row["medication_id"], row["schedule_key"]): row
        for row in existing
    }
    for row in existing:
        key = (row["medication_id"], row["schedule_key"])
        if row["status"] == "planned" and key not in desired_keys:
            con.execute("DELETE FROM dose_instances WHERE id=?", (row["id"],))

    medication_by_id = {med["id"]: med for med in scheduled}
    for item in desired:
        key = (item["medication_id"], item["schedule_key"])
        current = existing_by_key.get(key)
        if current is None:
            medication = medication_by_id[item["medication_id"]]
            con.execute(
                """
                INSERT OR IGNORE INTO dose_instances(
                    id,medication_id,person_id,scheduled_date,schedule_key,scheduled_time,
                    slot_label,dose_text,product_name_snapshot,ingredient_name_snapshot,status
                ) VALUES(?,?,?,?,?,?,?,?,?,?, 'planned')
                """,
                (
                    uuid_factory(), item["medication_id"], person_id, item["scheduled_date"],
                    item["schedule_key"], item["scheduled_time"], item["slot_label"], item["dose_text"],
                    medication["product_name"], medication.get("ingredient_name"),
                ),
            )
            continue
        if current["status"] != "planned":
            continue
        if (
            current["scheduled_time"] == item["scheduled_time"]
            and current["slot_label"] == item["slot_label"]
            and current["dose_text"] == item["dose_text"]
        ):
            continue
        con.execute(
            """
            UPDATE dose_instances
            SET scheduled_time=?, slot_label=?, dose_text=?
            WHERE id=? AND status='planned'
            """,
            (item["scheduled_time"], item["slot_label"], item["dose_text"], current["id"]),
        )

    rows = list(con.execute(
        """
        SELECT i.*,
               COALESCE(i.product_name_snapshot,m.product_name) AS product_name,
               COALESCE(i.ingredient_name_snapshot,m.ingredient_name) AS ingredient_name,
               m.meal_relation, m.administration_route
        FROM dose_instances i JOIN medications m ON m.id=i.medication_id
        WHERE i.person_id=? AND i.scheduled_date=? AND i.schedule_key NOT LIKE 'prn:%'
        ORDER BY CASE WHEN i.scheduled_time IS NULL THEN 1 ELSE 0 END,
                 i.scheduled_time, i.schedule_key, i.rowid
        """,
        (person_id, target.isoformat()),
    ).fetchall())
    rows.sort(key=lambda row: (
        (0, *clock_sort_key(row["scheduled_time"]))
        if row["scheduled_time"] is not None
        else (1, int(row["schedule_key"].split(":", 1)[1]), 0)
    ))
    return {
        "date": target.isoformat(),
        "doses": [dict(row) for row in rows],
        "prn_medications": prn,
        "unscheduled_medications": unscheduled,
        "summary": {
            "planned": sum(1 for row in rows if row["status"] == "planned"),
            "taken": sum(1 for row in rows if row["status"] == "taken"),
            "skipped": sum(1 for row in rows if row["status"] == "skipped"),
        },
    }


def record_instance(
    con: sqlite3.Connection,
    instance_id: str,
    status: str,
    occurred_at: str,
    note: str | None,
    uuid_factory: Callable[[], str],
    *,
    preserve_existing_same_state: bool = False,
) -> dict:
    row = con.execute("SELECT * FROM dose_instances WHERE id=?", (instance_id,)).fetchone()
    if row is None:
        raise KeyError("dose instance not found")

    existing_log = con.execute(
        "SELECT id,status,occurred_at,note FROM dose_logs WHERE dose_instance_id=?", (instance_id,)
    ).fetchone()
    if preserve_existing_same_state and row["status"] == status and existing_log is not None:
        return dict(row)

    con.execute(
        "UPDATE dose_instances SET status=?, completed_at=? WHERE id=?",
        (status, occurred_at, instance_id),
    )
    if existing_log is None:
        con.execute(
            """
            INSERT INTO dose_logs(
                id,medication_id,person_id,status,occurred_at,note,dose_instance_id,
                product_name_snapshot,dosage_text_snapshot
            ) VALUES(?,?,?,?,?,?,?,?,?)
            """,
            (
                uuid_factory(), row["medication_id"], row["person_id"], status, occurred_at,
                note, instance_id, row["product_name_snapshot"], row["dose_text"],
            ),
        )
    else:
        con.execute(
            "UPDATE dose_logs SET status=?, occurred_at=?, note=? WHERE id=?",
            (status, occurred_at, note, existing_log["id"]),
        )
    updated = con.execute("SELECT * FROM dose_instances WHERE id=?", (instance_id,)).fetchone()
    return dict(updated)


def cancel_instance_completion(con: sqlite3.Connection, instance_id: str) -> dict:
    row = con.execute("SELECT * FROM dose_instances WHERE id=?", (instance_id,)).fetchone()
    if row is None:
        raise KeyError("dose instance not found")
    con.execute("DELETE FROM dose_logs WHERE dose_instance_id=?", (instance_id,))
    if str(row["schedule_key"]).startswith("prn:"):
        # A PRN instance exists only because an actual intake was recorded. Undoing
        # that intake removes the ad-hoc occurrence instead of leaving an invisible
        # planned dose which can never appear in a fixed daily schedule.
        snapshot = dict(row)
        con.execute("DELETE FROM dose_instances WHERE id=?", (instance_id,))
        snapshot.update(status="canceled", completed_at=None, deleted=True)
        return snapshot
    con.execute(
        "UPDATE dose_instances SET status='planned', completed_at=NULL WHERE id=?",
        (instance_id,),
    )
    updated = con.execute("SELECT * FROM dose_instances WHERE id=?", (instance_id,)).fetchone()
    return dict(updated)


def create_prn_instance(
    con: sqlite3.Connection,
    medication: dict,
    occurred_date: date,
    uuid_factory: Callable[[], str],
) -> dict:
    if not medication.get("as_needed"):
        raise ValueError("medication is not PRN/as_needed")
    instance_id = uuid_factory()
    con.execute(
        """INSERT INTO dose_instances(
               id,medication_id,person_id,scheduled_date,schedule_key,scheduled_time,
               slot_label,dose_text,product_name_snapshot,ingredient_name_snapshot,status
           ) VALUES(?,?,?,?,?,NULL,?,?,?,?, 'planned')""",
        (
            instance_id,
            medication["id"],
            medication["person_id"],
            occurred_date.isoformat(),
            f"prn:{instance_id}",
            "필요시",
            medication.get("dosage_text"),
            medication.get("product_name"),
            medication.get("ingredient_name"),
        ),
    )
    return dict(con.execute("SELECT * FROM dose_instances WHERE id=?", (instance_id,)).fetchone())


def prn_taken_count(con: sqlite3.Connection, medication_id: str, target: date) -> int:
    rows = con.execute(
        """SELECT l.occurred_at FROM dose_logs l
           JOIN dose_instances i ON i.id=l.dose_instance_id
           WHERE l.medication_id=? AND l.status='taken' AND i.schedule_key LIKE 'prn:%'""",
        (medication_id,),
    ).fetchall()
    count = 0
    for row in rows:
        try:
            occurred = datetime.fromisoformat(str(row[0]))
            occurred_date = (
                occurred.astimezone(APP_TIMEZONE).date() if occurred.tzinfo else occurred.date()
            )
            if occurred_date == target:
                count += 1
        except ValueError:
            continue
    return count

File: rust/medicine_core/src/planning.rs
use chrono::{FixedOffset, NaiveDate, Utc};
use rusqlite::{params, Connection, TransactionBehavior};
use serde_json::{json, Value};
use std::path::Path;
use uuid::Uuid;

use crate::personal_db::{self, Access, OpenError};
use crate::planning_medications::{clock_sort_key, load_active_medications, Medication};

enum PlanningRoute<'a> {
    DailyPlan(&'a str),
}

pub(crate) enum PlanningError {
    BadRequest(String),
    NotFound(String),
    Unavailable,
    Internal,
}

pub(crate) struct DesiredDose {
    pub(crate) medication_id: String,
    pub(crate) schedule_key: String,
    pub(crate) scheduled_time: Option<String>,
    pub(crate) slot_label: Option<String>,
    pub(crate) dose_text: Option<String>,
}

struct ExistingDose {
    id: String,
    medication_id: String,
    schedule_key: String,
    status: String,
    scheduled_time: Option<String>,
    slot_label: Option<String>,
    dose_text: Option<String>,
}

pub(crate) fn handles_request(method: &str, path: &str) -> bool {
    route(method, path).is_some()
}

pub(crate) fn handle_request(
    personal_db: Option<&Path>,
    method: &str,
    raw_path: &str,
    path: &str,
) -> Option<(u16, Value)> {
    let route = route(method, path)?;
    let result = match route {
        PlanningRoute::DailyPlan(person_id) => daily_plan(personal_db, person_id, raw_path),
    };
    Some(match result {
        Ok(body) => (200, body),
        Err(PlanningError::BadRequest(detail)) => (400, json!({"detail": detail})),
        Err(PlanningError::NotFound(detail)) => (404, json!({"detail": detail})),
        Err(PlanningError::Unavailable) => {
            (503, json!({"detail": "personal database unavailable"}))
        }
        Err(PlanningError::Internal) => (500, json!({"detail": "unexpected server error"})),
    })
}

fn route<'a>(method: &str, path: &'a str) -> Option<PlanningRoute<'a>> {
    if !method.trim().eq_ignore_ascii_case("GET") {
        return None;
    }
    let person_id = path
        .strip_prefix("/api/people/")?
        .strip_suffix("/daily-plan")?;
    if person_id.is_empty() || person_id.contains('/') {
        return None;
    }
    Some(PlanningRoute::DailyPlan(person_id))
}

fn daily_plan(
    personal_db_path: Option<&Path>,
    person_id: &str,
    raw_path: &str,
) -> Result<Value, PlanningError> {
    let target = target_date(raw_path)?;
    let mut con = open_write(personal_db_path)?;
    let transaction = con
        .transaction_with_behavior(TransactionBehavior::Immediate)
        .map_err(|_| PlanningError::Internal)?;
    if !person_exists(&transaction, person_id)? {
        return Err(PlanningError::NotFound("person not found".to_owned()));
    }
    let medications = load_active_medications(&transaction, person_id, target)?;
    let plan = materialize(&transaction, person_id, &medications, target)?;
    transaction.commit().map_err(|_| PlanningError::Internal)?;
    Ok(plan)
}

pub(crate) fn materialize_for_dashboard(
    con: &Connection,
    person_id: &str,
    medications: &[Medication],
    target: NaiveDate,
) -> Result<Value, PlanningError> {
    materialize(con, person_id, medications, target)
}

pub(crate) fn target_date(raw_path: &str) -> Result<NaiveDate, PlanningError> {
    match query_parameter(raw_path, "date")? {
        Some(value) => parse_date(&value),
        None => Ok(today_kst()),
    }
}

fn open_write(personal_db_path: Option<&Path>) -> Result<Connection, PlanningError> {
    personal_db::open(personal_db_path, Access::ReadWrite).map_err(|error| match error {
        OpenError::Unavailable => PlanningError::Unavailable,
        OpenError::Sql => PlanningError::Internal,
    })
}

fn person_exists(con: &Connection, person_id: &str) -> Result<bool, PlanningError> {
    con.query_row(
        "SELECT EXISTS(SELECT 1 FROM people WHERE id=?)",
        [person_id],
        |row| row.get::<_, i64>(0),
    )
    .map(|exists| exists != 0)
    .map_err(|_| PlanningError::Internal)
}

fn materialize(
    con: &Connection,
    person_id: &str,
    medications: &[Medication],
    target: NaiveDate,
) -> Result<Value, PlanningError> {
    let applicable = medications
        .iter()
        .filter_map(|medication| match medication_applies(medication, target) {
            Ok(true) => Some(Ok(medication)),
            Ok(false) => None,
            Err(error) => Some(Err(error)),
        })
        .collect::<Result<Vec<_>, _>>()?;
    let prn = applicable
        .iter()
        .copied()
        .filter(|medication| medication.as_needed())
        .collect::<Vec<_>>();
    let scheduled = applicable
        .iter()
        .copied()
        .filter(|medication| !medication.as_needed())
        .collect::<Vec<_>>();
    let unscheduled = scheduled
        .iter()
        .copied()
        .filter(|medication| {
            medication.schedules.is_empty() && medication.frequency_per_day().unwrap_or(0) == 0
        })
        .collect::<Vec<_>>();
    let mut desired = Vec::new();
    for medication in &scheduled {
        desired.extend(desired_for(medication)?);
    }

    reconcile_instances(con, person_id, &scheduled, target, &desired)?;
    let doses = plan_doses(con, person_id, target)?;
    let summary = json!({
        "planned": doses.iter().filter(|dose| dose["status"] == "planned").count(),
        "taken": doses.iter().filter(|dose| dose["status"] == "taken").count(),
        "skipped": doses.iter().filter(|dose| dose["status"] == "skipped").count(),
    });
    Ok(json!({
        "date": target.format("%Y-%m-%d").to_string(),
        "doses": doses,
        "prn_medications": prn.into_iter().map(Medication::to_json).collect::<Vec<_>>(),
        "unscheduled_medications": unscheduled.into_iter().map(Medication::to_json).collect::<Vec<_>>(),
        "summary": summary,
    }))
}

pub(crate) fn medication_applies(
    medication: &Medication,
    target: NaiveDate,
) -> Result<bool, PlanningError> {
    if !medication.active() {
        return Ok(false);
    }
    if medication.start_date()?.is_some_and(|start| target < start) {
        return Ok(false);
    }
    if medication.end_date()?.is_some_and(|end| target > end) {
        return Ok(false);
    }
    Ok(true)
}

pub(crate) fn desired_for(medication: &Medication) -> Result<Vec<DesiredDose>, PlanningError> {
    let medication_id = medication.id()?.to_owned();
    if !medication.schedules.is_empty() {
        return medication
            .schedules
            .iter()
            .enumerate()
            .map(|(index, schedule)| {
                Ok(DesiredDose {
                    medication_id: medication_id.clone(),
                    schedule_key: format!("slot:{}", index + 1),
                    scheduled_time: Some(schedule.time_of_day.clone()),
                    slot_label: None,
                    dose_text: schedule
                        .dose_text
                        .clone()
                        .or_else(|| medication.dosage_text().map(str::to_owned)),
                })
            })
            .collect();
    }
    let frequency = medication.frequency_per_day().unwrap_or(0);
    if frequency <= 0 {
        return Ok(Vec::new());
    }
    Ok((1..=frequency)
        .map(|index| DesiredDose {
            medication_id: medication_id.clone(),
            schedule_key: format!("slot:{index}"),
            scheduled_time: None,
            slot_label: Some(format!("{index}회차")),
            dose_text: medication.dosage_text().map(str::to_owned),
        })
        .collect::<Vec<_>>())
}

fn reconcile_instances(
    con: &Connection,
    person_id: &str,
    scheduled: &[&Medication],
    target: NaiveDate,
    desired: &[DesiredDose],
) -> Result<(), PlanningError> {
    let target_text = target.format("%Y-%m-%d").to_string();
    let mut statement = con
        .prepare(
            "SELECT id,medication_id,schedule_key,status,scheduled_time,slot_label,dose_text
             FROM dose_instances WHERE person_id=? AND scheduled_date=?",
        )
        .map_err(|_| PlanningError::Internal)?;
    let existing = statement
        .query_map(params![person_id, target_text], |row| {
            Ok(ExistingDose {
                id: row.get(0)?,
                medication_id: row.get(1)?,
                schedule_key: row.get(2)?,
                status: row.get(3)?,
                scheduled_time: row.get(4)?,
                slot_label: row.get(5)?,
                dose_text: row.get(6)?,
            })
        })
        .map_err(|_| PlanningError::Internal)?
        .map(|row| row.map_err(|_| PlanningError::Internal))
        .collect::<Result<Vec<_>, _>>()?;

    for row in &existing {
        let desired_key = desired.iter().any(|item| {
            item.medication_id == row.medication_id && item.schedule_key == row.schedule_key
        });
        if row.status == "planned" && !desired_key {
            con.execute("DELETE FROM dose_instances WHERE id=?", [&row.id])
                .map_err(|_| PlanningError::Internal)?;
        }
    }

    for item in desired {
        let current = existing.iter().find(|row| {
            row.medication_id == item.medication_id && row.schedule_key == item.schedule_key
        });
        match current {
            None => {
                let medication = scheduled
                    .iter()
                    .copied()
                    .find(|medication| medication.id().ok() == Some(item.medication_id.as_str()))
                    .ok_or(PlanningError::Internal)?;
                con.execute(
                    "INSERT OR IGNORE INTO dose_instances(
                        id,medication_id,person_id,scheduled_date,schedule_key,scheduled_time,
                        slot_label,dose_text,product_name_snapshot,ingredient_name_snapshot,status
                     ) VALUES(?,?,?,?,?,?,?,?,?,?, 'planned')",
                    params![
                        Uuid::new_v4().to_string(),
                        item.medication_id,
                        person_id,
                        target_text,
                        item.schedule_key,
                        item.scheduled_time,
                        item.slot_label,
                        item.dose_text,
                        medication.product_name()?,
                        medication.ingredient_name(),
                    ],
                )
                .map_err(|_| PlanningError::Internal)?;
            }
            Some(current) if current.status != "planned" => {}
            Some(current)
                if current.scheduled_time == item.scheduled_time
                    && current.slot_label == item.slot_label
                    && current.dose_text == item.dose_text => {}
            Some(current) => {
                con.execute(
                    "UPDATE dose_instances SET scheduled_time=?,slot_label=?,dose_text=?
                     WHERE id=? AND status='planned'",
                    params![
                        item.scheduled_time,
                        item.slot_label,
                        item.dose_text,
                        current.id
                    ],
                )
                .map_err(|_| PlanningError::Internal)?;
            }
        }
    }
    Ok(())
}

fn plan_doses(
    con: &Connection,
    person_id: &str,
    target: NaiveDate,
) -> Result<Vec<Value>, PlanningError> {
    let target_text = target.format("%Y-%m-%d").to_string();
    let mut statement = con
        .prepare(
            "SELECT i.id,i.medication_id,i.person_id,i.scheduled_date,i.schedule_key,
                    i.scheduled_time,i.slot_label,i.dose_text,i.product_name_snapshot,
                    i.ingredient_name_snapshot,i.status,i.completed_at,i.created_at,
                    COALESCE(i.product_name_snapshot,m.product_name) AS product_name,
                    COALESCE(i.ingredient_name_snapshot,m.ingredient_name) AS ingredient_name,
                    m.meal_relation,m.administration_route
             FROM dose_instances i JOIN medications m ON m.id=i.medication_id
             WHERE i.person_id=? AND i.scheduled_date=? AND i.schedule_key NOT LIKE 'prn:%'",
        )
        .map_err(|_| PlanningError::Internal)?;
    let rows = statement
        .query_map(params![person_id, target_text], |row| {
            Ok(json!({
                "id": row.get::<_, String>(0)?,
                "medication_id": row.get::<_, String>(1)?,
                "person_id": row.get::<_, String>(2)?,
                "scheduled_date": row.get::<_, String>(3)?,
                "schedule_key": row.get::<_, String>(4)?,
                "scheduled_time": row.get::<_, Option<String>>(5)?,
                "slot_label": row.get::<_, Option<String>>(6)?,
                "dose_text": row.get::<_, Option<String>>(7)?,
                "product_name_snapshot": row.get::<_, Option<String>>(8)?,
                "ingredient_name_snapshot": row.get::<_, Option<String>>(9)?,
                "status": row.get::<_, String>(10)?,
                "completed_at": row.get::<_, Option<String>>(11)?,
                "created_at": row.get::<_, String>(12)?,
                "product_name": row.get::<_, String>(13)?,
                "ingredient_name": row.get::<_, Option<String>>(14)?,
                "meal_relation": row.get::<_, Option<String>>(15)?.unwrap_or_else(|| "unspecified".to_owned()),
                "administration_route": row.get::<_, Option<String>>(16)?.unwrap_or_else(|| "unknown".to_owned()),
            }))
        })
        .map_err(|_| PlanningError::Internal)?
        .map(|row| row.map_err(|_| PlanningError::Internal))
        .collect::<Result<Vec<_>, _>>()?;
    let mut keyed = rows
        .into_iter()
        .map(|dose| {
            let key = if let Some(time) = dose["scheduled_time"].as_str() {
                let (hour, minute) = clock_sort_key(time)?;
                (0_u32, hour, minute)
            } else {
                let slot = dose["schedule_key"]
                    .as_str()
                    .and_then(|key| key.split_once(':'))
                    .and_then(|(_, slot)| slot.parse::<u32>().ok())
                    .ok_or(PlanningError::Internal)?;
                (1_u32, slot, 0_u32)
            };
            Ok((key, dose))
        })
        .collect::<Result<Vec<_>, PlanningError>>()?;
    keyed.sort_by_key(|(key, _)| *key);
    Ok(keyed.into_iter().map(|(_, dose)| dose).collect())
}

fn parse_date(value: &str) -> Result<NaiveDate, PlanningError> {
    NaiveDate::parse_from_str(value, "%Y-%m-%d")
        .map_err(|_| PlanningError::BadRequest(format!("Invalid isoformat string: '{value}'")))
}

fn today_kst() -> NaiveDate {
    let korea = FixedOffset::east_opt(9 * 60 * 60).expect("KST offset is valid");
    Utc::now().with_timezone(&korea).date_naive()
}

fn query_parameter(raw_path: &str, key: &str) -> Result<Option<String>, PlanningError> {
    let Some((_, raw_query)) = raw_path.split_once('?') else {
        return Ok(None);
    };
    let query = raw_query
        .split_once('#')
        .map_or(raw_query, |(query, _)| query);
    let mut result = None;
    for pair in query.split('&') {
        let (raw_key, raw_value) = pair.split_once('=').unwrap_or((pair, ""));
        if percent_decode(raw_key)? != key {
            continue;
        }
        let value = percent_decode(raw_value)?;
        result = Some(value);
    }
    Ok(result)
}

fn percent_decode(value: &str) -> Result<String, PlanningError> {
    let bytes = value.as_bytes();
    let mut decoded = Vec::with_capacity(bytes.len());
    let mut index = 0;
    while index < bytes.len() {
        match bytes[index] {
            b'+' => decoded.push(b' '),
            b'%' if index + 2 < bytes.len() => {
                let high = hex_value(bytes[index + 1]).ok_or_else(|| {
                    PlanningError::BadRequest("invalid query encoding".to_owned())
                })?;
                let low = hex_value(bytes[index + 2]).ok_or_else(|| {
                    PlanningError::BadRequest("invalid query encoding".to_owned())
                })?;
                decoded.push((high << 4) | low);
                index += 2;
            }
            byte => decoded.push(byte),
        }
        index += 1;
    }
    String::from_utf8(decoded)
        .map_err(|_| PlanningError::BadRequest("invalid query encoding".to_owned()))
}

fn hex_value(value: u8) -> Option<u8> {
    match value {
        b'0'..=b'9' => Some(value - b'0'),
        b'a'..=b'f' => Some(value - b'a' + 10),
        b'A'..=b'F' => Some(value - b'A' + 10),
        _ => None,
    }
}
